mlir-sycl/lib/Dialect/SYCL/IR/SYCLTraits.cpp

```diff
@@ -16,152 +16,89 @@
 using namespace mlir;
 using namespace mlir::sycl;
 
-static mlir::LogicalResult
-verifyEqualDimensions(mlir::sycl::SYCLMethodOpInterface Op) {
-  const auto RetTy = Op->getResult(0).getType();
-  if (RetTy.isInteger(64)) {
-    return mlir::success();
+static LogicalResult verifyEqualDimensions(Operation *op) {
+  const auto retTy = op->getResult(0).getType();
+  if (retTy.isInteger(64)) {
+    return success();
   }
-  const unsigned ThisDimensions = getDimensions(Op.getBaseType());
-  const unsigned RetDimensions = getDimensions(RetTy);
-  if (ThisDimensions != RetDimensions) {
-    return Op->emitOpError("Base type and return type dimensions mismatch: ")
-           << ThisDimensions << " vs " << RetDimensions;
+  const unsigned thisDimensions = getDimensions(op->getOperandTypes()[0]);
+  const unsigned retDimensions = getDimensions(retTy);
+  if (thisDimensions != retDimensions) {
+    return op->emitOpError("base type and return type dimensions mismatch: ")
+           << thisDimensions << " vs " << retDimensions;
   }
-  return mlir::success();
+  return success();
 }
 
-mlir::LogicalResult mlir::sycl::verifySYCLGetComponentTrait(Operation *OpPtr) {
+LogicalResult mlir::sycl::verifySYCLGetComponentTrait(Operation *op) {
   // size_t get(int dimension) const;
   // size_t &operator[](int dimension);
   // size_t operator[](int dimension) const;
   // only available if Dimensions == 1
   // size_t operator size_t() const;
-  auto Op = cast<SYCLMethodOpInterface>(OpPtr);
-  const llvm::StringRef FunctionName = Op.getFunctionName();
-  const bool IsSizeTCast = Op.getFunctionName() == "operator unsigned long";
-  const mlir::Type RetTy = Op->getResult(0).getType();
-  const bool IsScalarReturn = RetTy.isIntOrIndex();
-  switch (Op->getNumOperands()) {
-  case 1: {
-    if (!IsSizeTCast) {
-      return Op->emitOpError("The ")
-             << FunctionName << " function expects an index argument";
-    }
-    if (!IsScalarReturn) {
-      return Op->emitOpError(
-                 "A cast to size_t must return a size_t value. Got ")
-             << RetTy;
-    }
-    const unsigned Dimensions = getDimensions(Op.getBaseType());
-    if (Dimensions != 1) {
-      return Op->emitOpError("A cast to size_t can only be performed when the "
-                             "number of dimensions is one. Got ")
-             << Dimensions;
-    }
-    break;
-  }
-  case 2: {
-    if (IsSizeTCast) {
-      return Op->emitOpError(
-          "A cast operation cannot recieve more than one argument");
-    }
-    if (FunctionName == "get" && !IsScalarReturn) {
-      return Op.emitOpError(
-          "The get method cannot return a reference, just a value");
-    }
-    break;
-  }
-  default:
-    llvm_unreachable("Invalid number of operands");
+  const auto resultTypes = op->getResultTypes();
+  const auto operandTypes = op->getOperandTypes();
+  if (operandTypes.size() == 1) {
```

Comment:
The code this is replacing had an `llvm_unreachable("Invalid number of operands");` for the case there are more than 2 operands. Is an assert(operandTypes.size() < 2 && "...") appropriate here?

Similar comment for the other code changes in this file.

Reply:
Good point. I added asserts to check that.